### ISSUE TITLE
improve connection handling and backend discovery

### ISSUE THREAD

**Issue by KodyKantor** [joyent/pgstatsmon#4]
pgstatsmon should handle issues that effect connections to Postgres:
* Postgres instances restarting
* Postgres instances disappearing for extended periods of time
* Postgres instances disappearing totally
* New Postgres instances appearing
* Dropped/corrupt/delayed packets

In order to address these, we can do a few things:
* Add query timeouts
* Add reconnect logic
* Add connection timeouts
* Add a Postgres discovery mechanism

Moray already handles query timeouts, so we can use a lot of that code (pulled into a separate PG client library) to handle the first point.

The second two bullets can be covered by using node-cueball. Cueball includes a 'connection set' object that maintains one connection per backend, which is exactly what we want for pgstatsmon. Tim Kordas did a lot of work getting Moray to use Cueball, so we can pull that code into the PG client library mentioned previously.

Cueball can use a 'resolver' which will inform cueball of when to add or remove connections. Ideally we would use the default DNS-based resolver for discovering our Postgres instances. Unfortunately our internal DNS configuration in Manta and Triton doesn't include SRV or A records for Postgres instances. Lookups for Postgres instances traditionally go directly to Zookeeper, which will return the cluster topology. Ideally pgstatsmon would remain unaware of Manatee, so this is an unfavorable option.

Alternatively, we could get information from SAPI about the deployed Postgres instances. We could funnel that data into the current pgstatsmon configuration file, possibly using something like config-agent. The config-agent optionally executes a process after a config file has been changed. We could set that action to send a signal to pgstatsmon to re-read its configuration file. We could then write a 'ConfigFileResolver' that informs the cueball ConnectionSet when instances are added or removed from the deployment based on the information read from the config file.

Another alternative is to improve the way Postgres instances register in DNS. I think the end result of this is most preferable, so we should investigate what sort of technical debt has accumulated here.

**Comment by KodyKantor** [comment]
Another idea I kicked around today was to implement a 'VMResolver' which uses VMAPI's changefeed support to keep a list of Postgres backends populated.

In the ConfigFileResolver case we will need something polling VMAPI to populate a config file. I would rather not have an out-of-process VMAPI poller. [RFD 5](https://github.com/joyent/rfd/blob/master/rfd/0005/README.md) addresses this by adding changefeed support to VMAPI.

It might be a better idea to implement a cueball Resolver based on VMAPI changefeed events instead of the ConfigFileResolver with a side-car VMAPI poller. I'll prototype this in the next couple days to see how it feels.

**Comment by KodyKantor** [comment]
After some discussion with the Manta team, we decided that since pgstatsmon can handle activities like backends disappearing for an indeterminate amount of time that we don't need VMAPI changefeed. Instead we designed a node-cueball resolver that occasionally polls VMAPI for information about deployed Postgres instances.

I did prototype the ConfigFileResolver, which worked, but having a sidecar process to perform backend discovery would have been a pain. It would introduce too many additional failure modes and general complication to the whole pgstatsmon stack without providing any benefit over the VMAPI cueball resolver.

The CR for this work is here: https://cr.joyent.us/#/c/3352/
